README.md
# To Run:

- Make sure the provider api is running
- npm install && npm run dev
- For testing npm run test


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.2.0](https://github.com/kami4u/test/compare/v1.1.0...v1.2.0) (2021-09-08)


### Features

* back ([1b9a950](https://github.com/kami4u/test/commit/1b9a950eae15052435a74deaf7ba0228f2fd28e3))

## [1.1.0](https://github.com/kami4u/test/compare/v1.0.7...v1.1.0) (2021-09-08)


### Features

* back ([c7a4904](https://github.com/kami4u/test/commit/c7a4904dbb977f731486df881bdc9e8822a5b572))
* prevent the application from crashing ([3497e98](https://github.com/kami4u/test/commit/3497e98d5b855284604357c83f380c03059d6ea3))

### [1.0.7](https://github.com/kami4u/test/compare/v1.0.6...v1.0.7) (2021-09-08)


### Bug Fixes

* prevent the application from crashing ([4f47416](https://github.com/kami4u/test/commit/4f47416d36fb95fd975be5f1c4beac8d1dc4da4d))

### [1.0.6](https://github.com/kami4u/test/compare/v1.0.5...v1.0.6) (2021-09-08)

### [1.0.5](https://github.com/kami4u/test/compare/v1.0.4...v1.0.5) (2021-09-08)

### [1.0.4](https://github.com/kami4u/test/compare/v1.0.3...v1.0.4) (2021-09-08)

### [1.0.3](https://github.com/kami4u/test/compare/v1.0.2...v1.0.3) (2021-09-08)

### [1.0.2](///compare/v1.0.1...v1.0.2) (2021-09-08)

### 1.0.1 (2021-09-08)


src/app.test.ts
import app from "./app";
import supertest from "supertest";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { gas, internal } from "./mockData/response";

const request = supertest(app);
const mock = new MockAdapter(axios);

describe("Provider", () => {
  it("Should give status 404 and error 'Payload is not right need to have name and callbackUrl'", async () => {
    const res = await request.post("/api/webhook");
    expect(res.status).toBe(404);
    expect(res.text).toBe("Payload is not right need to have name and callbackUrl");
  });

  it("Should give status 500 and error 'Provider Api is Failing with message #fail'", async () => {
    mock.onGet("http://127.0.0.1:3000/providers/gas").reply(500, "#fail");
    const res = await request.post("/api/webhook").send({ name: "gas", callbackUrl: "http://localhost:3100/api/data" });
    expect(res.status).toBe(200);
    expect(res.body.errors[0].statusCode).toBe(500);
    expect(res.body.errors[0].message).toBe("Provider Api is Failing with message #fail");
  });

  it("Should give status 404 and error 'Provider asd Does not exist'", async () => {
    mock.onGet("http://127.0.0.1:3000/providers/asd").reply(404, "#fail");
    const res = await request.post("/api/webhook").send({ name: "asd", callbackUrl: "http://localhost:3100/api/data" });
    expect(res.status).toBe(200);
    expect(res.body.errors[0].statusCode).toBe(404);
    expect(res.body.errors[0].message).toBe("Provider asd Does not exist");
  });

  it("Should give status 200 with the data", async () => {
    mock.onGet("http://127.0.0.1:3000/providers/gas").reply(200, gas);
    mock.onPost("http://localhost:3100/api/data").reply(200);
    const res = await request.post("/api/webhook").send({ name: "gas", callbackUrl: "http://localhost:3100/api/data" });
    expect(res.status).toBe(200);
    expect(res.body.providers[0].name).toBe("gas");
    expect(res.body.providers[0].data[0].data[0].amount).toBe(22.27);
  });

  it("Should give status 200 with the data containing all providers", async () => {
    mock.onGet("http://127.0.0.1:3000/providers/gas").reply(200, gas);
    mock.onGet("http://127.0.0.1:3000/providers/internal").reply(200, internal);
    mock.onPost("http://localhost:3100/api/data").reply(200);
    const res = await request
      .post("/api/webhook")
      .send({ name: ["gas", "internal"], callbackUrl: "http://localhost:3100/api/data" });

    expect(res.status).toBe(200);
    expect(res.body.providers[0].name).toBe("gas");
    expect(res.body.providers[1].name).toEqual("internal");
    expect(res.body.providers[0].data[0].data[0].amount).toEqual(22.27);
    expect(res.body.providers[1].data[0].data[0].amount).toEqual(15.12);
  });
});


src/app.ts
import express, { NextFunction, Request, Response } from "express";
import winston from "winston";
import cors from "cors";
import expressWinston from "express-winston";

import logger from "./utils/logger";
import routes from "./api/routes";

const app = express();

// request logging middleware
app.use(
  expressWinston.logger({
    transports: [new winston.transports.Console()],
    format: logger.format,
    meta: true, // optional: control whether you want to log the meta data about the request (default to true)
    msg: "{{res.statusCode}} {{req.method}} {{res.responseTime}}ms {{req.url}}", // optional: customize the default logging message. E.g. "{{res.statusCode}} {{req.method}} {{res.responseTime}}ms {{req.url}}"
    colorize: true, // Color the text and status code, using the Express/morgan color palette (text: gray, status: default green, 3XX cyan, 4XX yellow, 5XX red).
  }),
);

app.use(cors());

logger.initialise();

// main API modules
app.use("/api", routes);

// catch-all error handler middleware
app.use((err: any, req: Request, res: Response, next: NextFunction) => {
  winston.error(`${err.message}`);
  res.status(404).end(err.message);
});

export default app;


src/api/routes.ts
import { Request, Response, Router } from "express";
import axios from "axios";
import bodyParser from "body-parser";
import { asyncRoute } from "../utils/async";

const router = Router();
const jsonParser = bodyParser.json();

router.post(
  "/webhook",
  jsonParser,
  asyncRoute(async (req: Request, res: Response) => {
    const result = {
      providers: [] as any,
      errors: [] as any,
    };

    if (!req.body.name || !req.body.callbackUrl) {
      res.status(404).end("Payload is not right need to have name and callbackUrl");
      return;
    }

    let { name, callbackUrl } = req.body;

    if (!Array.isArray(name)) {
      name = [name];
    }

    for (const ele of name) {
      try {
        const response = await axios.get(`http://127.0.0.1:3000/providers/${ele}`);
        result.providers.push({ name: ele, data: response.data });
      } catch (error) {
        // Catch it if the server is not responding
        if (error.code === "ECONNREFUSED") {
          result.errors.push({
            name: ele,
            statusCode: 500,
            message: "Provider server is not responding",
          });
        } else if (error.response.status === 404) {
          result.errors.push({
            name: ele,
            statusCode: error.response.status,
            message: `Provider ${ele} Does not exist`,
          });
        } else if (error.response.status === 500) {
          result.errors.push({
            name: ele,
            statusCode: error.response.status,
            message: `Provider Api is Failing with message ${error.response.data}`,
          });
        }
      }
    }
    try {
      await axios.post(callbackUrl, { result });
    } catch (error) {
      result.errors.push({
        name: "callbackUrl",
        statusCode: 404,
        message: "Callback Url Does not Exist",
      });
    }
    res.send(result);
  }),
);

router.post(
  "/data",
  jsonParser,
  asyncRoute(async (req: Request, res: Response) => {
    res.end();
  }),
);

export default router;
